File: wasm-of-ocaml/benchmarks/C/arith_1000.c
// https://ocaml.org/learn/tutorials/99problems.html problems 32-34
// imperative implementation of the same problem. Some recursion replaced with iteration
int gcd(int a, int b){
    if (b == 0) return a;
    return gcd(b, a%b);
}

int coprime(int a, int b){
    return gcd(a, b) == 1;
}

int phi(int n){
    if (n == 1) return 1;
    int acc = 0;
    for (int d = 1; d < n; d++){
        if (coprime(n, d)) acc++;
    }
    return acc;
}

int main(){
    int x = 0;
    for (int i = 1000; i > 0; i--){
        x += phi(i);
    }
    return x; // so that above code can't be optimised away
}

File: wasm-of-ocaml/benchmarks/Ocaml/arith_1000.ml
(* https://ocaml.org/learn/tutorials/99problems.html problems 32-34 *)
let rec gcd a b =
    if b = 0 then a else gcd b (a mod b)
	
let coprime a b = gcd a b = 1

let phi n =
    let rec count_coprime acc d =
      if d < n then
        count_coprime (if coprime n d then acc + 1 else acc) (d + 1)
      else acc
    in
    if n = 1 then 1 else count_coprime 0 1

let rec stackphi = function
  | 0 -> 0
  | n -> (phi n) + stackphi (n-1)

let a = stackphi 1000
